rust/azure_iot_operations_connector/src/base_connector/managed_azure_device_registry.rs

```diff
@@ -835,10 +835,12 @@ impl AssetClient {
             Self::current_status_to_modify(&status_write_guard, updated_asset.version);
         let mut status_updated = false;
 
+        // make changes to a copy of the dataset_hashmap so that this function is cancel safe
+        let mut temp_dataset_hashmap = self.dataset_hashmap.clone();
         // update datasets
         // remove the datasets that are no longer present in the new asset definition.
         // This triggers deletion notification since this drops the update sender.
-        self.dataset_hashmap.retain(|dataset_name, _| {
+        temp_dataset_hashmap.retain(|dataset_name, _| {
```

Comment:
Just to confirm here, let's say we have dataset 1 here that will not be retained. I was thinking of a scenario as follows:
1. On creating a temporary dataset we now have two deletion tokens for dataset 1
2. When we retain on the temporary dataset we drop the copied deletion token which triggers a deletion on the dataset
3. Say we get cancelled 
4. On the second run we will once again create a temporary dataset for dataset 1 with two deletion tokens. 

On step 4, can we create two deletion tokens after one has been cancelled (I believe we use a drop guard so we lose it on step 2)?

Reply:
it's not a drop guard, but the sending side of the watch channel, so the temp dataset hashmap creates a copy of the sending side of the watch channel, so because the original still has a copy of the watch channel, it doesn't go out of scope and trigger deletion until the watch channel goes out of scope in both the real dataset_hashmap (right before the update is returned to the application) as well as the temp_dataset (in line 843)